### PR TITLE
feat(SubmitButton): The mission for submission GM-315

### PR DESCRIPTION
**Pull request by codecaaron**
### Goals
1. Standardize Submit Buttons by create a standalone RHF connected button that handles presentation logic for both button customization and loading states.
2. Expose more direct control of `loading` and `disabled` logic to reduce state duplication.
3. RHF Dependency Upgrade (for `isValidating`)
   
### Note to Reviewers
- Usage documentation in storybook is not included as the final access pattern is subject to change. 
- Currently the only access pattern this will be used in is `GridFrom` the standalone submit button is subject to change.

### Changes

* Creates standalone `SubmitButton` component.
* Adds Loading Spinner and `loading` prop.
* Adds new possible options for `loading` and `disabled` which accept callbacks for `disabled` and `loading` states be computed as a function of [`formState`](https://react-hook-form.com/v6/api#formState)
* Bumps RHF to `6.15.3` for access to `isValidating`.
* Fixes old version of gamut being added to the local modules via `@codecademy/monaco` (this was causing duplication of RHF)

### Usage

```tsx
<SubmitButton
  as={CTAButton}
  loading={({ isSubmitting }) => isSubmitting}
  disabled={({ isValid }) => !isValid} 
/>
```

You must now manually specify the conditions you would like to disable the submit button on, GridForm and SubmitButton will not assume a desired behavior.  A few reasons for this are:
- Disabled submit is largely presentational - submission triggers validation.
- Conditions are highly dependent on the context and adding default behaviors depending on configuration of GridForm will likely be very confusing.
- Most cases where we need disabled customization are due to computed state / external state (recaptcha) needing access.

Customizing either `loading` or `disabled` can be done with both a boolean or a function of [formState](https://react-hook-form.com/v6/api#formState)
1. Boolean: Passed Value
2. Callback: Replaces ALL logic for determining state and only uses the return value of the callback.

**Static Standalone**
```tsx
<SubmitButton 
  disabled={someBooleanValue || otherTing}
  loading={someBooleanValue && !otherThing}
>
    Submit
</SubmitButton>
```

**Computed Standalone**
```tsx
import { SubmitButton } from '@codecademy/gamut';

// Computed disabled state
<SubmitButton 
   disabled={({ isSubmitting, isValidating, isInvalid }) => isSubmitting || isValidating || isInvalid}
>
    Submit
</SubmitButton>

// Computed loading state
<SubmitButton 
   loading={({ isSubmitting, isValidating}) => isSubmitting || isValidating}
>
    Submit
</SubmitButton>
```

**Computed GridForm**
```tsx
import { GridForm } from '@codecademy/gamut';

// Computed disabled state
<GridForm 
  submit={{ 
    disabled:  ({ isSubmitting, isValidating, isInvalid }) => isSubmitting || isValidating || isInvalid,
 }}
/>


const MyComponent = ({ shutThisDown }) => {
  <GridForm 
     submit={{ 
    disabled:  ({ isSubmitting, isValidating, isInvalid }) => isSubmitting || isValidating || isInvalid || shutThisDown,
   }}
/>
}
```

### PR Checklist

- [x] Related to designs: Linked in JIRA
- [x] Related to JIRA ticket: https://codecademy.atlassian.net/browse/GM-315
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.2188fa.0<br/>@codecademy/gamut-labs@19.22.3-alpha.2188fa.0<br/>@codecademy/gamut@36.4.5-alpha.2188fa.0<br/>@codecademy/styleguide@43.7.5-alpha.2188fa.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611beb0f0dac5c1237dd75f0--gamut-preview.netlify.app>https://611beb0f0dac5c1237dd75f0--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611beb0f0dac5c1237dd75f0>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.cc0a3f.0<br/>@codecademy/gamut-labs@19.22.3-alpha.cc0a3f.0<br/>@codecademy/gamut@36.4.5-alpha.cc0a3f.0<br/>@codecademy/styleguide@43.7.5-alpha.cc0a3f.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611bf266cafcc30ff8f37c25--gamut-preview.netlify.app>https://611bf266cafcc30ff8f37c25--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611bf266cafcc30ff8f37c25>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.c6847f.0<br/>@codecademy/gamut-labs@19.22.3-alpha.c6847f.0<br/>@codecademy/gamut@36.4.5-alpha.c6847f.0<br/>@codecademy/styleguide@43.7.5-alpha.c6847f.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611c07a722c8b5008f88a1e4--gamut-preview.netlify.app>https://611c07a722c8b5008f88a1e4--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611c07a722c8b5008f88a1e4>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.ca69fc.0<br/>@codecademy/gamut-labs@19.22.3-alpha.ca69fc.0<br/>@codecademy/gamut@36.4.5-alpha.ca69fc.0<br/>@codecademy/monaco@0.2.1-alpha.ca69fc.0<br/>@codecademy/styleguide@43.7.5-alpha.ca69fc.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611c29cc94bc50188ebea552--gamut-preview.netlify.app>https://611c29cc94bc50188ebea552--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611c29cc94bc50188ebea552>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.9a5d61.0<br/>@codecademy/gamut-labs@19.22.3-alpha.9a5d61.0<br/>@codecademy/gamut@36.4.5-alpha.9a5d61.0<br/>@codecademy/monaco@0.2.1-alpha.9a5d61.0<br/>@codecademy/styleguide@43.7.5-alpha.9a5d61.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611c498f2bc7fb37c79d826d--gamut-preview.netlify.app>https://611c498f2bc7fb37c79d826d--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611c498f2bc7fb37c79d826d>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.0e8bc7.0<br/>@codecademy/gamut-labs@19.22.3-alpha.0e8bc7.0<br/>@codecademy/gamut@36.4.5-alpha.0e8bc7.0<br/>@codecademy/monaco@0.2.1-alpha.0e8bc7.0<br/>@codecademy/styleguide@43.7.5-alpha.0e8bc7.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611c6e2208a24a6c46bcd7ab--gamut-preview.netlify.app>https://611c6e2208a24a6c46bcd7ab--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611c6e2208a24a6c46bcd7ab>Deploy Logs</a></p>

**Review by dreamwasp — APPROVED** [review]
perfection 💖

**Review by vonbarown — APPROVED** [review]

**Review by jenesh — APPROVED** [review]
Button works great, just gotta fix console.warn errors for unit tests

```bash
  console.warn
    📋 `formState.isValid` is applicable with `onTouched`, `onChange` or `onBlur` mode. https://react-hook-form.com/api#formState

      at Object.isValid [as get] (node_modules/react-hook-form/src/useForm.ts:1284:25)
      at SubmitButton (node_modules/@codecademy/gamut/dist/Form/SubmitButton.js:23:39)
      at renderWithHooks (node_modules/react-dom/cjs/react-dom.development.js:14985:18)
      at updateFunctionComponent (node_modules/react-dom/cjs/react-dom.development.js:17356:20)
      at beginWork (node_modules/react-dom/cjs/react-dom.development.js:19063:16)
      at beginWork$1 (node_modules/react-dom/cjs/react-dom.development.js:23940:14)
      at performUnitOfWork (node_modules/react-dom/cjs/react-dom.development.js:22779:12)
```

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.8a5b64.0<br/>@codecademy/gamut-labs@19.22.3-alpha.8a5b64.0<br/>@codecademy/gamut@36.4.5-alpha.8a5b64.0<br/>@codecademy/monaco@0.2.1-alpha.8a5b64.0<br/>@codecademy/styleguide@43.7.5-alpha.8a5b64.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611e9c84bf67344141c1dcc5--gamut-preview.netlify.app>https://611e9c84bf67344141c1dcc5--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611e9c84bf67344141c1dcc5>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.37-alpha.dd4c06.0<br/>@codecademy/gamut-labs@19.22.3-alpha.dd4c06.0<br/>@codecademy/gamut@36.4.5-alpha.dd4c06.0<br/>@codecademy/monaco@0.2.1-alpha.dd4c06.0<br/>@codecademy/styleguide@43.7.5-alpha.dd4c06.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611e9fca4a407b3c5ea38293--gamut-preview.netlify.app>https://611e9fca4a407b3c5ea38293--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611e9fca4a407b3c5ea38293>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.40-alpha.e0622d.0<br/>@codecademy/gamut-labs@19.23.3-alpha.e0622d.0<br/>@codecademy/gamut@36.4.7-alpha.e0622d.0<br/>@codecademy/monaco@0.2.1-alpha.e0622d.0<br/>@codecademy/styleguide@43.7.9-alpha.e0622d.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611ea3c86903c747329fa0af--gamut-preview.netlify.app>https://611ea3c86903c747329fa0af--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611ea3c86903c747329fa0af>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.40-alpha.0b0569.0<br/>@codecademy/gamut-labs@19.23.3-alpha.0b0569.0<br/>@codecademy/gamut@36.4.7-alpha.0b0569.0<br/>@codecademy/monaco@0.2.1-alpha.0b0569.0<br/>@codecademy/styleguide@43.7.9-alpha.0b0569.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611ed424743a73719270c722--gamut-preview.netlify.app>https://611ed424743a73719270c722--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611ed424743a73719270c722>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.40-alpha.6a52f1.0<br/>@codecademy/gamut-labs@19.23.3-alpha.6a52f1.0<br/>@codecademy/gamut@36.4.7-alpha.6a52f1.0<br/>@codecademy/monaco@0.2.1-alpha.6a52f1.0<br/>@codecademy/styleguide@43.7.9-alpha.6a52f1.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611edbc29d61346f313fcf6e--gamut-preview.netlify.app>https://611edbc29d61346f313fcf6e--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611edbc29d61346f313fcf6e>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.40-alpha.d43fbd.0<br/>@codecademy/gamut-labs@19.23.3-alpha.d43fbd.0<br/>@codecademy/gamut@36.4.7-alpha.d43fbd.0<br/>@codecademy/monaco@0.2.1-alpha.d43fbd.0<br/>@codecademy/styleguide@43.7.9-alpha.d43fbd.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611f05b6078200b974896208--gamut-preview.netlify.app>https://611f05b6078200b974896208--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611f05b6078200b974896208>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.40-alpha.4fee23.0<br/>@codecademy/gamut-labs@19.23.3-alpha.4fee23.0<br/>@codecademy/gamut@36.4.7-alpha.4fee23.0<br/>@codecademy/monaco@0.2.1-alpha.4fee23.0<br/>@codecademy/styleguide@43.7.9-alpha.4fee23.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611f16904a407bc208a382ec--gamut-preview.netlify.app>https://611f16904a407bc208a382ec--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611f16904a407bc208a382ec>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.42-alpha.31c625.0<br/>@codecademy/gamut-labs@19.23.5-alpha.31c625.0<br/>@codecademy/gamut@36.5.1-alpha.31c625.0<br/>@codecademy/monaco@0.2.1-alpha.31c625.0<br/>@codecademy/styleguide@43.8.1-alpha.31c625.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611fc3dc9d613450d33fce7c--gamut-preview.netlify.app>https://611fc3dc9d613450d33fce7c--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611fc3dc9d613450d33fce7c>Deploy Logs</a></p>

**Review by dreamwasp — APPROVED** [review]
this all looks ✨, the computed state are especially *chef's kiss*

File: packages/gamut/src/Form/SubmitButton.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  loading = false,
+  ...rest
+}) => {
+  const { formState } = useFormContext();
+
+  const isLoading =
+    typeof loading === 'function' ? loading(formState) : loading;
+
+  const isDisabled =
+    typeof disabled === 'function' ? disabled(formState) : disabled;
```
Comment (JoshuaKGoldberg) [inline packages/gamut/src/Form/SubmitButton.tsx:31]: Nit: could move this down below the `if` to not call it unnecessarily?

File: packages/styleguide/stories/Organisms/GridForm.stories.mdx [hunk 2]
```diff
@@ -1105,3 +1106,52 @@ If you need to checkboxes to fit into a smaller space, you can use our our `tigh
     )}
   </Story>
 </Canvas>
+
+## Loading
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Organisms/GridForm.stories.mdx:1110]: I don't understand the second example. Can you add some explanatory text here?

File: packages/monaco/package.json [hunk 3]
```diff
@@ -14,8 +14,8 @@
     "react-dom": ">=16.8.1"
   },
   "dependencies": {
-    "@codecademy/gamut": "35.2.0",
-    "@codecademy/gamut-styles": "11.6.2",
+    "@codecademy/gamut": "^36.4.4",
+    "@codecademy/gamut-styles": "^11.9.1",
```
Comment (JoshuaKGoldberg) [inline packages/monaco/package.json:18]: Related/unrelated?

**Review by JoshuaKGoldberg — APPROVED** [review]
Very nice! Glad we finally have an official loading button state 🎉

File: packages/monaco/package.json (hunk 3 above)
Comment (codecaaron) [reply packages/monaco/package.json:18]: Basically it was causing 2 versions of RHF to be installed erroneously from an old gamut version.  This was causing duped versions while installing, I made a note of it in the description.

Comment (JoshuaKGoldberg) [reply packages/monaco/package.json:18]: OH I misinterpreted, thanks haha

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.43-alpha.93c84b.0<br/>@codecademy/gamut-labs@19.23.6-alpha.93c84b.0<br/>@codecademy/gamut@36.6.1-alpha.93c84b.0<br/>@codecademy/monaco@0.2.1-alpha.93c84b.0<br/>@codecademy/styleguide@43.8.2-alpha.93c84b.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://61200c8634001b00afa44c47--gamut-preview.netlify.app>https://61200c8634001b00afa44c47--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/61200c8634001b00afa44c47>Deploy Logs</a></p>

**Pull request merged**